Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,526 @@
+        }
+        value = UInt64(self[0])
+        for digit in self[1...] {
+            value *= 10
+            value += UInt64(digit)
+        }
+        return value
+    }
+}
+
+internal struct UInt128: Equatable, Hashable {
+    /// The high order 64 bits
+    internal var hi: UInt64
+    /// The low order 64 bits
+    internal var lo: UInt64
+
+    internal init(hi: UInt64, lo: UInt64) {
+        self.hi = hi
+        self.lo = lo
+    }
+
+    internal init() {
+        self.hi = 0
+        self.lo = 0
+    }
+
+    internal static func multiply(_ left: UInt64, by right: UInt64) -> UInt128 {
+        let (product, didOverflow) = left.multipliedReportingOverflow(by: right)
+        guard didOverflow else {
+            // no overflow means we can just use our regular multiplication result
+            return UInt128(hi: 0, lo: product)
+        }
+
+        let leftHi = left.upper32bits
+        let leftLo = left.lower32bits
+
+        let rightHi = right.upper32bits
+        let rightLo = right.lower32bits
+
+        var productHi = leftHi * rightHi
+        let productMid0 = leftHi * rightLo
+        let productMid1 = leftLo * rightHi
+        var productLo = leftLo * rightLo
+
+        productHi += productMid0.upper32bits
+        let productPartial = productMid0.lower32bits + productMid1 + productLo.upper32bits
+
+        productHi += productPartial.upper32bits
+        productLo = (productPartial << 32) + productLo.lower32bits
+
+        return UInt128(hi: productHi, lo: productLo)
+    }
+
+    internal func divideBy1Billion() -> (quotient: UInt128, remainder: Int) {
+        // swiftlint:disable:previous cyclomatic_complexity
+        let denominator: UInt64 = 1000 * 1000 * 1000
+        var remainder: UInt64 = 0
+        var quotient = self
+
+        guard !(quotient.hi == 0 && quotient.lo == 0) else {
+            return (quotient: quotient, remainder: 0)
+        }
+
+        for i in 0...3 {
+            // Adjust remainder to match value of next dividend
+            remainder <<= 32
+            // Add the divided to remainder
+            var quotientI: UInt64
+            switch i {
+            case 0: quotientI = quotient.hi.upper32bits
+            case 1: quotientI = quotient.hi.lower32bits
+            case 2: quotientI = quotient.lo.upper32bits
+            case 3: quotientI = quotient.lo.lower32bits
+            default: quotientI = 0
+            }
+            remainder += quotientI
+            // quotient[i] = Int(remainder / DIVISOR)
+            switch i {
+            case 0: quotient.hi = (((remainder / denominator) << 32) | quotient.hi.lower32bits)
+            case 1: quotient.hi = ((remainder / denominator).lower32bits | quotient.hi.getBitsUnshifted(0...31))
+            case 2: quotient.lo = (((remainder / denominator) << 32) | quotient.lo.lower32bits)
+            case 3: quotient.lo = ((remainder / denominator).lower32bits | quotient.lo.getBitsUnshifted(0...31))
+            default: _ = ()
```

Comment:
nit: I think the preferred way to no-op in a case is with `break`; see [here](https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html#ID139)